File: src/main/java/xyz/devvydont/smprpg/items/blueprints/resources/slayer/NecroticFleshFamilyBlueprint.java
package xyz.devvydont.smprpg.items.blueprints.resources.slayer;

import io.papermc.paper.datacomponent.item.Consumable;
import io.papermc.paper.datacomponent.item.consumable.ConsumeEffect;
import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;
import org.jspecify.annotations.NonNull;
import xyz.devvydont.smprpg.items.CustomItemType;
import xyz.devvydont.smprpg.items.ItemClassification;
import xyz.devvydont.smprpg.items.base.CustomCompressableBlueprint;
import xyz.devvydont.smprpg.items.interfaces.IEdible;
import xyz.devvydont.smprpg.items.interfaces.IConsumable;
import xyz.devvydont.smprpg.services.ItemService;
import xyz.devvydont.smprpg.util.crafting.CompressionRecipeMember;
import xyz.devvydont.smprpg.util.crafting.MaterialWrapper;
import xyz.devvydont.smprpg.util.time.TickTime;

import java.util.ArrayList;
import java.util.List;

public class NecroticFleshFamilyBlueprint extends CustomCompressableBlueprint implements IEdible, IConsumable {

    public static final List<CompressionRecipeMember> COMPRESSION_FLOW = List.of(
            new CompressionRecipeMember(new MaterialWrapper(CustomItemType.NECROTIC_FLESH)),
            new CompressionRecipeMember(new MaterialWrapper(CustomItemType.PREMIUM_NECROTIC_FLESH)),
            new CompressionRecipeMember(new MaterialWrapper(CustomItemType.ENCHANTED_NECROTIC_FLESH)),
            new CompressionRecipeMember(new MaterialWrapper(CustomItemType.NECROTIC_FLESH_SINGULARITY))
    );

    public NecroticFleshFamilyBlueprint(ItemService itemService, CustomItemType type) {
        super(itemService, type);
    }

    @Override
    public List<CompressionRecipeMember> getCompressionFlow() {
        return COMPRESSION_FLOW;
    }

    @Override
    public int getNutrition(ItemStack item) {
        return 1;
    }

    @Override
    public float getSaturation(ItemStack item) {
        return 1;
    }

    @Override
    public boolean canAlwaysEat(ItemStack item) {
        return true;
    }

    @Override
    public @NonNull Consumable getConsumableComponent(ItemStack item) {

        var effects = new ArrayList<ConsumeEffect>();

        if (getCustomItemType().equals(CustomItemType.NECROTIC_FLESH_SINGULARITY))
            effects.add(ConsumeEffect.applyStatusEffects(List.of(
                    new PotionEffect(PotionEffectType.WITHER, (int) TickTime.minutes(5), 9),
                    new PotionEffect(PotionEffectType.BLINDNESS, (int) TickTime.minutes(5), 1),
                    new PotionEffect(PotionEffectType.DARKNESS, (int) TickTime.minutes(5), 1)
            ), 1f));

        if (getCustomItemType().equals(CustomItemType.ENCHANTED_NECROTIC_FLESH))
            effects.add(ConsumeEffect.applyStatusEffects(List.of(
                    new PotionEffect(PotionEffectType.WITHER, (int) TickTime.minutes(5), 4),
                    new PotionEffect(PotionEffectType.BLINDNESS, (int) TickTime.minutes(5), 0)
            ), 1f));

        if (getCustomItemType().equals(CustomItemType.PREMIUM_NECROTIC_FLESH))
            effects.add(ConsumeEffect.applyStatusEffects(List.of(
                    new PotionEffect(PotionEffectType.WITHER, (int) TickTime.minutes(5), 2)
            ), 1f));

        return Consumable.consumable()
                .consumeSeconds(10)
                .addEffects(effects)
                .build();
    }

    @Override
    public ItemClassification getItemClassification() {
        return ItemClassification.CONSUMABLE;
    }

}

File: src/main/java/xyz/devvydont/smprpg/items/blueprints/resources/slayer/NecroticFleshFamilyBlueprint.kt
package xyz.devvydont.smprpg.items.blueprints.resources.slayer

import io.papermc.paper.datacomponent.item.Consumable
import io.papermc.paper.datacomponent.item.consumable.ConsumeEffect
import net.kyori.adventure.key.Key
import org.bukkit.inventory.ItemStack
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType
import xyz.devvydont.smprpg.items.CustomItemType
import xyz.devvydont.smprpg.items.ItemClassification
import xyz.devvydont.smprpg.items.base.CustomItemBlueprint
import xyz.devvydont.smprpg.items.interfaces.ICompressible
import xyz.devvydont.smprpg.items.interfaces.ICompressible.CompressionStep
import xyz.devvydont.smprpg.items.interfaces.IConsumable
import xyz.devvydont.smprpg.items.interfaces.IEdible
import xyz.devvydont.smprpg.items.interfaces.IModelOverridden
import xyz.devvydont.smprpg.items.interfaces.ISellable
import xyz.devvydont.smprpg.services.ItemService
import xyz.devvydont.smprpg.util.extensions.calculateCompressedWorth
import xyz.devvydont.smprpg.util.time.TickTime

class NecroticFleshFamilyBlueprint(itemService: ItemService, type: CustomItemType) :
    CustomItemBlueprint(itemService, type), ICompressible, ISellable, IEdible, IConsumable, IModelOverridden {

    override val itemClassification: ItemClassification get() = ItemClassification.CONSUMABLE

    override val decompressor: CompressionStep? get() = when (customItemType) {
        CustomItemType.PREMIUM_NECROTIC_FLESH -> CompressionStep(itemService.getBlueprint(CustomItemType.NECROTIC_FLESH) as ICompressible, 1, 9)
        CustomItemType.ENCHANTED_NECROTIC_FLESH -> CompressionStep(itemService.getBlueprint(CustomItemType.PREMIUM_NECROTIC_FLESH) as ICompressible, 1, 9)
        CustomItemType.NECROTIC_FLESH_SINGULARITY -> CompressionStep(itemService.getBlueprint(CustomItemType.ENCHANTED_NECROTIC_FLESH) as ICompressible, 1, 9)
        else -> null
    }

    override val compressor: CompressionStep? get() = when (customItemType) {
        CustomItemType.NECROTIC_FLESH -> CompressionStep(itemService.getBlueprint(CustomItemType.PREMIUM_NECROTIC_FLESH) as ICompressible, 9, 1)
        CustomItemType.PREMIUM_NECROTIC_FLESH -> CompressionStep(itemService.getBlueprint(CustomItemType.ENCHANTED_NECROTIC_FLESH) as ICompressible, 9, 1)
        CustomItemType.ENCHANTED_NECROTIC_FLESH -> CompressionStep(itemService.getBlueprint(CustomItemType.NECROTIC_FLESH_SINGULARITY) as ICompressible, 9, 1)
        else -> null
    }

    override fun getWorth(itemStack: ItemStack): Int = calculateCompressedWorth(itemStack)

    override fun getNutrition(item: ItemStack?): Int = 0
    override fun getSaturation(item: ItemStack?): Float = 0.0f
    override fun canAlwaysEat(item: ItemStack?): Boolean = true

    override fun getConsumableComponent(item: ItemStack?): Consumable {
        val effects = ArrayList<ConsumeEffect>()

        if (customItemType == CustomItemType.NECROTIC_FLESH_SINGULARITY) effects.add(
            ConsumeEffect.applyStatusEffects(
                listOf(
                    PotionEffect(PotionEffectType.WITHER, TickTime.minutes(5).toInt(), 9),
                    PotionEffect(PotionEffectType.BLINDNESS, TickTime.minutes(5).toInt(), 1),
                    PotionEffect(PotionEffectType.DARKNESS, TickTime.minutes(5).toInt(), 1)
                ), 1f
            )
        )

        if (customItemType == CustomItemType.ENCHANTED_NECROTIC_FLESH) effects.add(
            ConsumeEffect.applyStatusEffects(
                listOf(
                    PotionEffect(PotionEffectType.WITHER, TickTime.minutes(5).toInt(), 4),
                    PotionEffect(PotionEffectType.BLINDNESS, TickTime.minutes(5).toInt(), 0)
                ), 1f
            )
        )

        if (customItemType == CustomItemType.PREMIUM_NECROTIC_FLESH) effects.add(
            ConsumeEffect.applyStatusEffects(
                listOf(
                    PotionEffect(PotionEffectType.WITHER, TickTime.minutes(5).toInt(), 2)
                ), 1f
            )
        )

        return Consumable.consumable()
            .consumeSeconds(10f)
            .addEffects(effects)
            .build()
    }

    override fun getDisplayKey(): Key? {
        return IModelOverridden.ofItemTypeInDirectory(CustomItemType.NECROTIC_FLESH, "materials")
    }
}
